### ISSUE TITLE
What's next?

### ISSUE THREAD

**Issue by eugenpar** [harveybc/gym-forex#4]
Hi! How can I use the calculated results from "neat-checkpoint-199"?

**Comment by harveybc** [comment]
Hi!, the genomes are saved with picke.dump, you must be able to load them with pickle.load and then evaluate them with some data in a similar way that is made during evolution.

BUT you will have a problem using it as in the example, since it overfits terribly, that is why you should use a separated feature extractor that reduce dimensionality trained with techniques that reduce overfitting such as dropout or batch normalization, and using its output to feed a neural network (controller) that can be evolved (using the feature extractor output as training data). 

Sorry for the lack of documentation or more complete examples and tests. I am working on implementing tests, some methodology and proper documentation, and making the projects more modular starting from preprocessing, feture-engineering , feature extraction (autoencoder-based) and evolved neural network as controller, but it will take a month or two if i dont die from the big C.

Please be patient.

**Comment by eugenpar** [comment]
Harvey, thanks for the prompt response. I'll start digging in the indicated direction, 
but I am not sure of any immediate success, since it is problematic to understand 
the data structure without describing it.
It would be nice to test your developments on me as a test site.
I'll have questions.
And let's not talk about sad things-God save us.
    Evgeniy.

**Comment by harveybc** [comment]
Evgeniy, sorry for the shallow response, i used [neat-python](https://neat-python.readthedocs.io/en/latest/index.html), you should use [restore_checkpoint ](https://neat-python.readthedocs.io/en/latest/_modules/checkpoint.html) methosd to load the population contained in the checkpoint,  so you can evaluate them in a way similar how is done in [agents/genome-evaluator.py](https://github.com/harveybc/gym-fx/blob/master/agents/genome_evaluator.py) methods evaluate_genome() (line 90) where the neural networks are created from the genomes and simulate() (line 65) wich activate the networks with data to get their outputs.

I hope you have success.

If you want to help me with testing, lets start from the beginning,  please start with the repo: [preprocessor](https://github.com/harveybc/preprocessor), it has some sub modules that are already pasing the tests, but you can check if the tests are right and if the documentation is ok.

I apologyze for my lack of order in existing repos. But If you want to helpme, i am refactoring the whole project (p2p neat with fx-trading application), i have divided it in many projects with reusable submodules. i am applying the test-driven methodology, so i first make the user manual(like requirements) then i made the tests, then i fix my existing code to pass the tests.

The idea is to be able to use each component independently via command-line (it can help for automation). 

The  repos i am refactoring are: 

1. preprocessing: data-trimming, standardization, sliding-window, feature-selection
2. feaure-engineering: (euristic training signal generator, MSSA decomposer, MSSA predictor 
3. feature-extractor: keras 1D deep conv net / LSTM trainer, autoencoder trainer, autoencoder_encoder evaluator.
4. gym-forex : forex trading account simulation environment compatible with OpenAI gym
5. gym-agent: agent  that uses keras  feature-extractor with dense layer as controller, another agent but  with neural network evolved with neat as controller.
6. Optimizer: uses neat to optimize a neural network using feature-extractor output as training data.
7. Singularity: extends neat to a decentralized arhitecture for collaboration during optimization.

I have descriptions, parameters and tests descriptions for each one, but since i am just starting whit the refactoring, it will probably take some weeks. 

Let me know any questions.

**Comment by eugenpar** [comment]
I will work through Your message in detail as a guide to action.
It may be more convenient to use your mailbox at times. My address pomos51@yandex.ru

**Comment by harveybc** [comment]
My address is harveybc@ingeni-us.com

**Comment by jjphung** [comment]
May I get added as a participant also please? I have refactoring, clean up, tidying up changes in my local branch that I think would be beneficial to this project.

**Comment by wilsonssss** [comment]
@jjphung bravo!

**Comment by harveybc** [comment]
@jjphung i will check tour local branch. But right now i am integrating my repo:harveybc/data-logger to store optimization  configurations  and results and visualize them, and control everything via Web. But after that, my plans are to totally remake the gym-forex repo with a plugin architecture (for agents, env configs) and a decent testing methodology. Thanks for your work, i've been distracted last weeks reading  the book "a thousand brains", amazing stuff! But Will try to use My scarse free time again in my proyects .

**Comment by jjphung** [comment]
@harveybc Thank you for the work you've done so far and will do! That book looks very interesting. I will try to get my local branch pushed up ASAP so you can review it in a PR. But I think I need the github permissions to do so (Settings -> Collaborators). 

Let me know if you need any help with features/testing outside of ML/RL/AI. I feel like you are the go-to expert here for that, but I think I can be of help with features around testing, system design, documentation, and basic AI features. It's really interesting to see someone with a TDD approach as you've mentioned in the above comment.

Edit: Feature branch ready to push up for review, just need edit permissions to this repo. `remote: Permission to harveybc/gym-fx.git denied to jjphung.`
